### ISSUE TITLE
DiffPIR leads to NaN values

### ISSUE THREAD

**Issue by jscanvic** [deepinv/deepinv#222]
Depending on the RNG state, DiffPIR may silently return a tensor ridden with NaN entries.

**Reproduction steps**

```python
import torch
from torch.nn.functional import interpolate
from deepinv.sampling import DiffPIR
from deepinv.models import DRUNet
from deepinv.optim import L2
from deepinv.physics import Downsampling, LinearPhysics, GaussianNoise


class Downsampling(LinearPhysics):
    def __init__(self, factor):
        super().__init__()
        self.factor = factor

    def A(self, x):
        return interpolate(x, scale_factor=1 / self.factor, mode="bicubic")

    def A_adjoint(self, y):
        return interpolate(y, scale_factor=self.factor, mode="bicubic")


torch.manual_seed(1)

physics = Downsampling(factor=3)
physics.noise_model = GaussianNoise(sigma=5/255)

denoiser = DRUNet(pretrained="download")
model = DiffPIR(model=denoiser, data_fidelity=L2())

x = torch.randn(1,3,32,32)
y = physics(x)

torch.manual_seed(2)

x_hat = model(y, physics)
assert not torch.isnan(x_hat).any()

torch.manual_seed(3)

x_hat = model(y, physics)
assert not torch.isnan(x_hat).any()
```

Version of deepinv: 0.2.0

**Observed behavior**

The first assert statement does not fail, but the second one does.

**Expected behavior**

Both assert statements should not fail.

**Comment by jscanvic** [comment]
After a bit of debugging, it appears that the NaN values are introduced by the function `deepinv.optim.utils.conjugate_gradient` which is used in `LinearPhysics.prox_l2`. This happens because the conjugate gradient algorithm may diverge as `physics.A_adjoint` is not the true adjoint of `physics.A`, which is **not** a bug in the library but in the user-provided class `Downsampling`.

Nonetheless, I suggest adding a divergence check in `conjugate_gradient` to let the user know why it failed, instead of silently introducing NaN values.

For reference, here is the snippet I've used to make sure all NaN values come from `conjugate_gradient`.

```python
import torch
from torch.nn.functional import interpolate
from deepinv.sampling import DiffPIR
from deepinv.models import DRUNet
from deepinv.optim import L2
from deepinv.physics import Downsampling, LinearPhysics, GaussianNoise


# a dummy conjugate gradient implementation
def conjugate_gradient(A, b, max_iter=1e2, tol=1e-5, eps=1e-8):
    return b


class Downsampling(LinearPhysics):
    def __init__(self, factor):
        super().__init__()
        self.factor = factor

    def A(self, x):
        return interpolate(x, scale_factor=1 / self.factor, mode="bicubic")

    def A_adjoint(self, y):
        return interpolate(y, scale_factor=self.factor, mode="bicubic")

    # same as LinearPhysics.prox_l2 but using the dummy conjugate gradient implementation
    def prox_l2(self, z, y, gamma):
        b = self.A_adjoint(y) + 1 / gamma * z
        H = lambda x: self.A_adjoint(self.A(x)) + 1 / gamma * x
        x = conjugate_gradient(H, b, self.max_iter, self.tol)
        return x

torch.manual_seed(1)

physics = Downsampling(factor=3)
physics.noise_model = GaussianNoise(sigma=5/255)

denoiser = DRUNet(pretrained="download")
model = DiffPIR(model=denoiser, data_fidelity=L2())

x = torch.randn(1,3,32,32)
y = physics(x)

torch.manual_seed(2)

x_hat = model(y, physics)
assert not torch.isnan(x_hat).any()

torch.manual_seed(3)

x_hat = model(y, physics)
# does not fail anymore thanks to the dummy conjugate gradient implementation
assert not torch.isnan(x_hat).any()
```
